runtime/src/iree/hal/drivers/cuda/tracing.c

```diff
@@ -16,8 +16,33 @@
 // To prevent spilling pages we leave some room for the context structure.
 #define IREE_HAL_CUDA_TRACING_DEFAULT_QUERY_CAPACITY (16 * 1024 - 256)
 
+// iree_hal_cuda_tracing_context_event_t contains a hipEvent that is used to
+// record timestamps. This event also belongs to 2 linked list.
+//
+// --------------------->---Submissions--->----------
+// \                     \                    \
+//  \                     \                    \
+// command_buffer        command_buffer          command_buffer
+//
+// The higher level list is owned by the tracing context and
+// elements are inserted and removed as commmand_buffers are
+// submitted and when they complete. This is a list of the head
+// elements for each command buffer.
+// The second list is the list of events within a command_buffer
```

Comment:
"The commnad buffer list is owned by the command buffer. It is the list of events used to trace command buffer dispatches.

When the event is in the freelist, next_submission should be null, and we reuse next_in_command_buffer to track the next free event."